src/components/ActionSheetAwareScrollView/ActionSheetKeyboardSpace.tsx

```diff
@@ -0,0 +1,411 @@
+    UNKNOWN: 0,
+    OPENING: 1,
+    OPEN: 2,
+    CLOSING: 3,
+    CLOSED: 4
+}
+const useAnimatedKeyboard = () => {
+    const state = useSharedValue(KeyboardState.UNKNOWN);
+    const height = useSharedValue(0);
+    const progress = useSharedValue(0);
+    const heightWhenOpened = useSharedValue(0);
+
+    useKeyboardHandler({
+        onStart: (e) => {
+            "worklet";
+
+            // save the last keyboard height
+            if (e.height === 0) {
+                heightWhenOpened.value = height.value;
+            }
+
+            // console.log("onStart", e, new Date().getTime());
+
+            if (e.height > 0) {
+                state.value = KeyboardState.OPENING;
+            } else {
+                state.value = KeyboardState.CLOSING;
+            }
+        },
+        onMove: (e) => {
+            "worklet";
+
+            // console.log("onMove", e, new Date().getTime());
+
+            progress.value = e.progress;
+            height.value = e.height;
+        },
+        onEnd: (e) => {
+            "worklet";
+
+            // console.log("onEnd", e, new Date().getTime());
+
+            if (e.height > 0) {
+                state.value = KeyboardState.OPEN;
+            } else {
+                state.value = KeyboardState.CLOSED;
+            }
+
+            height.value = e.height;
+            progress.value = e.progress;
+        },
+    }, []);
+
+    return { state, height, heightWhenOpened, progress };
+};
+const setInitialValueAndRunAnimation = (value: number, animation: number) => {
+    "worklet";
+
+    return withSequence(
+        withTiming(value, { duration: 0 }),
+        animation,
+    )
+}
+
+const useSafeAreaPaddings = () => {
+    const StyleUtils = useStyleUtils();
+    const insets = useSafeAreaInsets();
+    const {paddingTop, paddingBottom} = StyleUtils.getSafeAreaPadding(insets ?? undefined);
+
+    return { top: paddingTop, bottom: paddingBottom };
+};
+
+const config = {
+    mass: 3,
+    stiffness: 1000,
+    damping: 500,
+};
+
+function ActionSheetKeyboardSpace(props: ViewProps) {
+    const styles = useThemeStyles();
+    const safeArea = useSafeAreaPaddings();
+    const keyboard = useAnimatedKeyboard();
+
+    // similar to using `global` in worklet but it's just a local object
+    const [syncLocalWorkletState] = useState({
+        lastState: KeyboardState.UNKNOWN,
+    });
+    const {windowHeight} = useWindowDimensions();
+    const {currentActionSheetState, transitionActionSheetStateWorklet: transition, transitionActionSheetState, resetStateMachine} = useContext(ActionSheetAwareScrollViewContext);
+
+    // Reset state machine when component unmounts
+    useEffect(() => () => resetStateMachine(), [resetStateMachine]);
+
+    useAnimatedReaction(
+        () => keyboard.state.value,
+        (lastState) => {
+            if (lastState === syncLocalWorkletState.lastState) {
+                return;
+            }
+
+            syncLocalWorkletState.lastState = lastState;
```

Comment:
Don't typically see this done for state – is this meant to re-render the component by potentially changing the state?

Reply:
@roryabraham see my answer above - I'll resolve this thread to keep a communication in single place if you don't mind (because anyway we're talking about the same thing).